symfony/symfony
# [FrameworkBundle] Functional testing system refactoring

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | tiny (for tests)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14590
| License       | MIT
| Doc PR        | no

Please refer to initial discussions here: #14590. 

There is a overview of the changes in a nutshell:

- Fix fetching of the test kernel class from the KERNEL_DIR that [returned only basename of the first file ends with *Kernel](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php#L128-L133) and not actually the class name and therefore could not be instantiated with `new` operator (probably because of this it was neccesary to [completely override getKernelClass](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/SecurityBundle/Tests/Functional/WebTestCase.php#L36-L41) to explicitly specify kernel name FQCN)
- Add ability to specify explicitly test kernel FQCN on the phpunit configuration level to avoid kinds of collision (please see **Problem** section of #14590)
- Add ability to override test kernel environment and debug-mode on the phpunit configuration level for example to support integration testing with real application kernel but in different environments.
- Introduce TestKernelInterface and extract [FrameworkBundle's test kernel class](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/AppKernel.php) in TestKernelInterface's implementation with adoptation to be usefull in arbitrary bundles.
- Remove code duplication and scattered temp directories clean in FrameworkBundle, SecurityBundle, TwigBundle (moved to TestKernel usage)
